Avoid guard to expr conversions

romainbrenguier opened the pull request:
For complicated programs, guards can get big and when represented as BDDs the conversion to exprt can be expensive. The goal of this pull request is to avoid these conversions, in particular in the solver we can convert directly from BDDs to literals without having to go through exprt first.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Make SSA_step guard be an optionalt<guardt>

This makes it explicit that it may not be set (previously was nil), and
having a guardt better represents what the guard is. In the case where
guard_exprt is used for guards this makes not difference in terms of
what data is used (it just restricts the interface to operations that
are meaningful on guards), in case where BDDs are used for guards, this
avoids a translation from BDDs to exprt, which can be expensive.

romainbrenguier committed:
Do not merge cond_expr

romainbrenguier committed:
Add bdd strategy for transforming bdd_exprt

This allows transformation on BDDs that are different from the standard
transformation to exprt.
This in particular will be used to convert BDDs directly to literals
without first going through exprts

romainbrenguier committed:
Use strategyt to convert BDDs to solver literals

This is more efficient that converting BDDs to expr and then
performing the conversion on the generated expr.

romainbrenguier committed:
Add a version of assignment with guardt argument

This can be used to avoid conveting guardt to exprt and back to guardt.

romainbrenguier committed:
Use version of assignment with guardt argument

This avoids some guardt to exprt conversion and back.

romainbrenguier committed:
guardt argument for symex_target::location

This avoids conversion to expr then back to guardt

romainbrenguier committed:
guardt argument for symex_target::decl

Avoid conversion from guardt to exprt and back.

romainbrenguier committed:
guardt argument for symex_target_equationt::goto_instruction

Avoids conversion from guardt to exprt and back

romainbrenguier committed:
Avoid usage of guardt::as_expr in symex_assign_symbol

The conversion can be expensive when the guards are big, so it is better
to avoid it.

romainbrenguier committed:
guardt argument for symex_target_equationt::function_call

Avoid conversion from guardt to exprt and back

romainbrenguier committed:
guardt argument for symex_target_equationt::shared_read

romainbrenguier committed:
guardt argument for symex_target_equationt::shared_write

romainbrenguier committed:
guardt argument for symex_target_equationt::dead

romainbrenguier committed:
guardt argument for symex_target_equationt::function_return

romainbrenguier committed:
guardt argument for symex_target_equationt::assumption

romainbrenguier committed:
guardt argument for symex_target_equationt::assertion

owen-mc-diffblue commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -590,8 +590,6 @@ void symex_target_equationt::convert_io(
 /// \param SSA_step The step you want to have shared values.
 void symex_target_equationt::merge_ireps(SSA_stept &SSA_step)
 {
-  merge_irep(SSA_step.guard);
```
This should be in the next commit, I think

owen-mc-diffblue commented on src/solvers/prop/bdd_expr.h:
```diff
@@ -36,6 +37,110 @@ class bdd_exprt
+    std::function<returnt(const returnt &, const returnt &)> make_or;
+    std::function<returnt(const returnt &, const returnt &)> make_and;
+    std::function<returnt(const returnt &)> make_not;
+    std::function<returnt(const returnt &, const returnt &, const returnt)> make_ite;
```
I don't understand `ite`. After a minute I guessed "if then else". Perhaps make it clearer somehow.

owen-mc-diffblue commented on src/solvers/prop/prop_conv_solver.cpp:
```diff
@@ -584,3 +585,21 @@ std::string prop_conv_solvert::decision_procedure_text() const
+      return prop.lor(prop.land(i, t), prop.land(!i, e));
+  };
+  return guard.apply(strategy);
+}
```
new line missing

owen-mc-diffblue reviewed (commented):
I got up to " Use version of assignment with guardt argument "

kroening commented:
This PR may set a new record in terms of the number of labels.

owen-mc-diffblue commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -486,8 +486,9 @@ void goto_symext::symex_assign_symbol(
-    make_and(state.guard.as_expr(), conjunction(guard));
+  // Temporarily add the state guard
+  guardt assignment_guard = state.guard;
+  assign_guard.add(conjunction(guard));
```
`assign_guard` -> `assignment_guard`?

owen-mc-diffblue commented on src/goto-symex/symex_atomic_section.cpp:
```diff
@@ -57,11 +57,9 @@ void goto_symext::symex_atomic_end(statet &state)
         ++it)
       read_guard|=*it;
-    exprt read_guard_expr=read_guard.as_expr();
-    do_simplify(read_guard_expr);
```
Is it a functional change that we no longer simplify `read_guard` here?

smowton replied:
I suspect we ought to have a simplify(guardt &) method, which simplifies when `guardt` == `exprt` but is a no-op on a BDD, which is always inherently in simplified form

owen-mc-diffblue commented on src/goto-symex/symex_atomic_section.cpp:
```diff
@@ -77,11 +77,9 @@ void goto_symext::symex_atomic_end(statet &state)
         ++it)
       write_guard|=*it;
-    exprt write_guard_expr=write_guard.as_expr();
-    do_simplify(write_guard_expr);
```
Is it a functional change that we no longer simplify `write_guard` here?

smowton commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -29,7 +29,7 @@ void symex_target_equationt::shared_read(
   SSA_stept &SSA_step=SSA_steps.back();
 
-  SSA_step.guard=guard;
+  SSA_step.guard.make_true().add(guard);
```
Given all these uses of `make_true().add(...)`, perhaps add a utility method `set` or `replace`?

smowton commented on src/solvers/prop/bdd_expr.h:
```diff
@@ -36,6 +37,110 @@ class bdd_exprt
+      auto insert_result = cache.emplace(r.id(), leaf(false_exprt()));
+      if(insert_result.second)
+      {
+        insert_result.first->second = [&]() -> returnt {
```
Suggest this big function deserves a name, rather than being anonymous

smowton commented on src/solvers/prop/bdd_expr.h:
```diff
@@ -36,6 +37,110 @@ class bdd_exprt
+  template <typename returnt>
+  struct strategyt
+  {
+    std::function<returnt(const exprt &)> leaf;
```
Why `std::function` everywhere, instead of virtual functions?

smowton commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -381,7 +381,7 @@ void symex_target_equationt::convert_guards(
         mstream << messaget::eom;
```
Commit message: convertion -> conversion

smowton commented on src/goto-symex/symex_target.h:
```diff
@@ -12,6 +12,7 @@ Author: Daniel Kroening, kroening@kroening.com
 #ifndef CPROVER_GOTO_SYMEX_SYMEX_TARGET_H
 #define CPROVER_GOTO_SYMEX_SYMEX_TARGET_H
```
Commit message: conveting -> converting

smowton commented:
Oh damn, one of the labels was "do not review". Oh well.

Re: `strategyt`, I suspect actually given how it's used we should simply use normal non-virtual functions, i.e. something like

```
template<resultt> struct x {
  resultt f();
  resultt g();
}

struct y : public x<int> {
}

int y::f() { }
int y::g() { }
```

Then any function that uses it (just `apply` I think) is templated?

smowton commented:
As it stands we get a weird mix of templates and dynamic dispatch, which I don't know to be wrong but feels... novel.

romainbrenguier committed:
Declare symex equations after guard_manager

The SSA step could contains BDDs which reference the guard_manager, they
must thus be destroyed before the guard_manager is, so their declaration
should appear after it.
This is required for the next intended step which is to make guards in
equations be of type guardt.

romainbrenguier committed:
Mark overriding functions as override

romainbrenguier committed:
Don't merge guard/cond_expr if already simplified

When the guards are represented by BDDs, the result is already using
sharing so merging here can be a waste of time.

romainbrenguier committed:
Fix condition for simplification

romainbrenguier committed:
Handle SSA_step.cond_expr directly to the solver

Without first converting to exprt.

romainbrenguier committed:
Add underlying() methods on guard implementations

This allow to retrieve the underlying implementations used by guards
which is necessary if the user of guards wants access the guard in a way
that was not intended in the interface.
The usage we want to make of this is in the handling of guards to the
solver, we need to know whether the guard is an exprt or a bddt in order
to make this transformation as efficient as possible.

romainbrenguier committed:
Clarify comment about trivial condition in decl

Explicit that this condition is there for the solver.

romainbrenguier committed:
Move convert_guard up [TO SQUASH]

romainbrenguier committed:
Use guardt::implies instead of guard_expr

This way we will be able to give a guardt to the target.assumption
method, which avoids some conversion to expr when using BDDs.

romainbrenguier committed:
Make copy of not_exprt private

With this copy constructor, we can accidentaly make a copy of the
expression instead of negating it, for instance:

not_exprt a{b}; // a = not b
not_exprt c{(expr)a}; // c = not (not b)
not_exprt d{a}; // d = not b !!!

romainbrenguier committed:
Use convert_guard in convert_decls

Avoid converting to expr before giving the formula to the solver.
This has little impact as the result is ignored anyway.

romainbrenguier committed:
Call convert guard to avoid guard to expr conversion

romainbrenguier committed:
Make SSA_step.cond_expr be an optional guardt

This avoids translation of guards to expression, which in the case of
BDDs can be wasteful as the translation to the solver will be more
efficient on BDDs than on the resulting expression.

romainbrenguier committed:
Special case of then == not else in BDD translation

This can make the resulting expression smaller by avoiding duplication
of the branches, for instance we can have
`a = b = c` instead of `a ? (b ? c : !c) : !(b ? c : !c)`

romainbrenguier committed:
Add implies method to guards

This should ultimately replace guard_expr

romainbrenguier committed:
Relax order in vcc tests

The order within expressions may change, in particular when using BDDs
for guards.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: db7ae45732de31ff559a9c57b4535a606959625c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/124790348
